I get a "module failure" when i tried to delete a second time (by mistake) a host in the icinga2 monitoring using icinga2_host module :
```
 Playbook run took 0 days, 0 hours, 1 minutes, 18 seconds
╭─cedric@cedric-VirtualBox ~/Documents/ansible  ‹release/2.x› 
╰─$ ansible-playbook -v playbooks/test.yml -e @~/vars.txt -u ansible --become

Using ~/Documents/ansible/ansible.cfg as config file

PLAY [INV***] **************************************************************************************************************************************************************************************************************************

TASK [SETUP | Gathering Facts] **************************************************************************************************************************************************************************************************************
ok: [INV***] (0.298394s)

PLAY [PRV***] **************************************************************************************************************************************************************************************************************************

TASK [SETUP | Gathering Facts] **************************************************************************************************************************************************************************************************************
ok: [PRV***] (0.899171s)

TASK [ICINGA2_HOST | Deploy icinga2 host] ***************************************************************************************************************************************************************************************************
failed: [PRV***] (item=INV***) => {
    "changed": false, 
    "failed": true, 
    "item": "INV***", 
    "module_stderr": "Warning: Identity file /var/lib/rundeck/.ssh/id_rsa not accessible: No such file or directory.
    Traceback (most recent call last):
      File \"/tmp/ansible_bpSVlp/ansible_module_icinga2_host.py\", line 341, in <module>
        main()
      File \"/tmp/ansible_bpSVlp/ansible_module_icinga2_host.py\", line 327, in main
        elif icinga.diff(name, data):
      File \"/tmp/ansible_bpSVlp/ansible_module_icinga2_host.py\", line 217, in diff
        ic_data = ret['data']['results'][0]
    TypeError: string indices must be integers, not str
    ", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE", 
    "rc": 1
}
```

My playbook is : 
```
- hosts: "INV***"
- hosts: "PRV***"
  become: True
  tasks:
    - name: "Deploy icinga2 host"
      icinga2_host:
        url: "https://172.***:5665"
        url_username: "icinga_api"
        url_password: "icinga"
        state: "absent"
        name: "{{ item }}"
        display_name: "{{ item }}"
        ip: "{{ hostvars[item]['ansible_default_ipv4']['address'] }}"
        zone: "{{ hostvars[item]['dc_domain_name'] }}"
        validate_certs: False
        variables: '{{ hostvars | hostvars_to_icinga2_host([item], groups) }}'
      with_items:
        - "INV***"
```

I've add an error message in the module icinga2_host.py, and it gets a "404 not found" error when it interrrogates the icinga2 server.

--- a/lib/ansible/modules/monitoring/icinga2_host.py
+++ b/lib/ansible/modules/monitoring/icinga2_host.py
@@ -92,7 +92,7 @@ EXAMPLES = '''
 - name: Add host to icinga
   icinga_host:
     icinga_server: "icinga.example.com"
-    icinga_user: "anisble
+    icinga_user: "anisble"
     icinga_pass: "mypassword"
     state: present
     name: "{{ansible_fqdn }}"
@@ -116,7 +116,7 @@ from ansible.module_utils.urls import fetch_url, url_argument_spec
 class icinga2_api:
     module = None
 
-    def call_url(self, path, data={}, method='GET'):
+    def call_url(self, path, data='', method='GET'):
         headers = {
             'Accept': 'application/json',
             'X-HTTP-Method-Override': method,
@@ -125,9 +125,9 @@ class icinga2_api:
         rsp, info = fetch_url(module=self.module, url=url, data=data, headers=headers, method=method)
         body = ''
         if rsp:
-          body = json.loads(rsp.read())
+            body = json.loads(rsp.read())
         if info['status'] >= 400:
-          body = info['body']
+            {body = info['body']
         return {'code': info['status'], 'data': body}
 
     def check_connection(self):
@@ -206,7 +206,7 @@ def main():
         ip=dict(required=True),
         variables=dict(type='dict', default=None),
     )
-    
+
     # Define the main module
     module = AnsibleModule(
         argument_spec=argument_spec,
@@ -305,4 +305,3 @@ def main():
 # import module snippets
 if __name__ == '__main__':
     main()
-

--- a/lib/ansible/modules/monitoring/icinga2_host.py
+++ b/lib/ansible/modules/monitoring/icinga2_host.py
@@ -104,6 +104,8 @@ import json
 import warnings
 warnings.simplefilter('ignore', requests.packages.urllib3.exceptions.SecurityWarning)
 
+from ansible.module_utils.basic import AnsibleModule
+
 
 # ===========================================
 # Icinga2 API class
@@ -133,7 +135,7 @@ class icinga2_api:
     def check_connection(self):
         ret = self.call_url('v1/status')
         if ret['code'] == 200:
-           return True
+            return True
         return False
 
     def exists(self, hostname):
@@ -142,8 +144,8 @@ class icinga2_api:
         }
         ret = self.call_url("v1/objects/hosts", data)
         if ret['code'] == 200:
-           if len(ret['data']['results']) == 1:
-               return True
+            if len(ret['data']['results']) == 1:
+                return True
         return False
 
     def create(self, hostname, data):
@@ -155,7 +157,7 @@ class icinga2_api:
         return ret
 
     def delete(self, hostname):
-        data = { "cascade": 1 }
+        data = {"cascade": 1}
         ret = self.call_url(
             url="v1/objects/hosts/" + hostname,
             data=data,
@@ -221,9 +223,9 @@ def main():
     name = module.params["name"]
     zone = module.params["zone"]
     template = []
-    template.append (name)
+    template.append(name)
     if module.params["template"]:
-        template.append (module.params["template"])
+        template.append(module.params["template"])
     check_command = module.params["check_command"]
     ip = module.params["ip"]
     display_name = module.params["display_name"]

--- a/lib/ansible/modules/monitoring/icinga2_host.py
+++ b/lib/ansible/modules/monitoring/icinga2_host.py
@@ -117,6 +117,7 @@ import json
 import warnings
 warnings.simplefilter('ignore', requests.packages.urllib3.exceptions.SecurityWarning)
 
+
 # ===========================================
 # Icinga2 API class
 #
@@ -127,12 +128,12 @@ class icinga2_api:
     password = None
     ssl_ca = None
 
-    def call_url (self, url, data = {}, methode = 'GET' ):
+    def call_url(self, url, data={}, methode='GET'):
         full_url = "https://" + self.server + ":" + str(self.port) + "/" + url
         headers = {
-                'Accept': 'application/json',
-                'X-HTTP-Method-Override': methode,
-                }
+            'Accept': 'application/json',
+            'X-HTTP-Method-Override': methode,
+        }
         r = requests.post(
             full_url,
             headers=headers,
@@ -140,7 +141,7 @@ class icinga2_api:
             data=json.dumps(data),
             verify=self.ssl_ca
         )
-        return { 'code': r.status_code, 'data': r.json() }
+        return {'code': r.status_code, 'data': r.json()}
 
     def check_connection(self):
         ret = self.call_url('v1/status')
@@ -148,44 +149,44 @@ class icinga2_api:
            return True
         return False
 
-    def exists (self, hostname):
+    def exists(self, hostname):
         data = {
-          "filter":  "match(\"" + hostname + "\", host.name)",
+            "filter": "match(\"" + hostname + "\", host.name)",
         }
-        ret = self.call_url ("v1/objects/hosts", data)
+        ret = self.call_url("v1/objects/hosts", data)
         if ret['code'] == 200:
            if len(ret['data']['results']) == 1:
-             return True
+               return True
         return False
 
-    def create (self, hostname, data):
-        ret = self.call_url (
-            url="v1/objects/hosts/"+ hostname,
+    def create(self, hostname, data):
+        ret = self.call_url(
+            url="v1/objects/hosts/" + hostname,
             data=data,
             methode="PUT"
         )
         return ret
 
-    def delete (self, hostname):
+    def delete(self, hostname):
         data = { "cascade": 1 }
-        ret = self.call_url (
-            url="v1/objects/hosts/"+ hostname,
+        ret = self.call_url(
+            url="v1/objects/hosts/" + hostname,
             data=data,
             methode="DELETE"
         )
         return ret
 
-    def modify (self, hostname, data):
-        ret = self.call_url (
-            url="v1/objects/hosts/"+ hostname,
+    def modify(self, hostname, data):
+        ret = self.call_url(
+            url="v1/objects/hosts/" + hostname,
             data=data,
             methode="POST"
         )
         return ret
 
-    def diff (self, hostname, data):
-        ret = self.call_url (
-            url="v1/objects/hosts/"+ hostname,
+    def diff(self, hostname, data):
+        ret = self.call_url(
+            url="v1/objects/hosts/" + hostname,
             data={},
             methode="GET"
         )
@@ -198,12 +199,13 @@ class icinga2_api:
                 changed = True
         return changed
 
+
 # ===========================================
 # Module execution.
 #
 def main():
     module = AnsibleModule(
-        argument_spec = dict(
+        argument_spec=dict(
             server=dict(required=True),
             port=dict(default=5665, type='int'),
             user=dict(default=None),
@@ -231,15 +233,15 @@ def main():
     state = module.params["state"]
     name = module.params["name"]
     zone = module.params["zone"]
-    template=[]
+    template = []
     template.append (name)
     if module.params["template"]:
-        template.append(module.params["template"])
+        template.append (module.params["template"])
     check_command = module.params["check_command"]
     ip = module.params["ip"]
     display_name = module.params["display_name"]
     if not display_name:
-      display_name = name
+        display_name = name
     variables = module.params["variables"]
 
     if not os.path.exists(ssl_ca):
@@ -290,7 +292,7 @@ def main():
         elif icinga.diff(name, data):
             if module.check_mode:
                 module.exit_json(changed=False, name=name, data=data)
-            ret = icinga.modify(name,data)
+            ret = icinga.modify(name, data)
             if ret['code'] == 200:
                 changed = True
             else:
@@ -309,13 +311,12 @@ def main():
                         changed = True
                     else:
                         module.fail_json(msg="bad return code creating host: %s" % (ret['data']))
-                except Exception:
-                    e = get_exception()
+                except Exception as e:
                     module.fail_json(msg="exception creating host: " + str(e))
         elif icinga.diff(name, data):
             if module.check_mode:
                 module.exit_json(changed=False, name=name, data=data)
-            ret = icinga.modify(name,data)
+            ret = icinga.modify(name, data)
             if ret['code'] == 200:
                 changed = True
             else:

